### PR TITLE
Add custom transports tests

### PR DESCRIPTION
**Pull request by pablogs9**
TODO:
- [x] Revert agent tag when https://github.com/eProsima/Micro-XRCE-DDS-Agent/pull/205 merged (see comment)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -31,7 +31,7 @@ project(microxrcedds VERSION "2.0.0")
 
 set(_client_tag develop)
 set(_client_version 1.2.5)
-set(_agent_tag develop)
+set(_agent_tag feature/custom_agent)
```
Comment (pablogs9) [inline CMakeLists.txt:34]: Commit when https://github.com/eProsima/Micro-XRCE-DDS-Agent/pull/205 merged

```suggestion
set(_agent_tag develop)
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+            auto data = client_to_agent_packet_queue[index].front();
+            client_to_agent_packet_queue[index].pop();
+
+            if (data.size() <= buffer_length)
+            {
+                std::copy(data.begin(), data.end(), buffer);
+                rv = data.size();
+                received = true;
+                printf("Custom agent receive: %ld in queue %d\n", rv, index);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:68]:
```suggestion
                std::cout << "Custom agent receive: " << rv << " bytes in queue " <<  index << std::endl;
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+{
+    return true;
+};
+
+eprosima::uxr::CustomAgent::RecvMsgFunction agent_custom_transport_read_packet = [&](
+            eprosima::uxr::CustomEndPoint* source_endpoint,
+            uint8_t* buffer,
+            size_t buffer_length,
+            int timeout,
+            eprosima::uxr::TransportRc& transport_rc) -> ssize_t
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:45]:
```suggestion
        eprosima::uxr::CustomEndPoint* source_endpoint,
        uint8_t* buffer,
        size_t buffer_length,
        int timeout,
        eprosima::uxr::TransportRc& transport_rc) -> ssize_t
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+static std::map<int32_t, stream_fifo> client_to_agent_stream_queue;
+static std::map<int32_t, stream_fifo> agent_to_client_stream_queue;
+
+std::mutex client_to_agent_mtx;
+std::mutex agent_to_client_mtx;
+
+template <class T>
+static int32_t find_queue_with_data(std::map<int32_t, T> m)
+{
+    for (auto const& it : m) {
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:20]:
```suggestion
    for (auto const& it : m)
    {
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        size_t message_length,
+        eprosima::uxr::TransportRc& transport_rc) -> ssize_t
+{
+    std::unique_lock<std::mutex> lock(agent_to_client_mtx);
+    int32_t index = static_cast<int32_t>(destination_endpoint->get_member<uint32_t>("index"));
+
+    std::vector<uint8_t> packet(buffer, buffer + message_length);
+    agent_to_client_packet_queue[index].emplace(std::move(packet));
+    transport_rc = eprosima::uxr::TransportRc::ok;
+    printf("Custom agent send: %ld\n", message_length);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:100]:
```suggestion
    std::cout << "Custom agent send: " << message_length << " bytes." << std::endl;
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+    return static_cast<ssize_t>(message_length);
+};
+
+eprosima::uxr::CustomAgent::RecvMsgFunction agent_custom_transport_read_stream = [&](
+            eprosima::uxr::CustomEndPoint* source_endpoint,
+            uint8_t* buffer,
+            size_t buffer_length,
+            int timeout,
+            eprosima::uxr::TransportRc& transport_rc) -> ssize_t
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:110]:
```suggestion
        eprosima::uxr::CustomEndPoint* source_endpoint,
        uint8_t* buffer,
        size_t buffer_length,
        int timeout,
        eprosima::uxr::TransportRc& transport_rc) -> ssize_t
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                               client_to_agent_stream_queue[index].size() :
+                               buffer_length;
+
+            for (size_t i = 0; i < rv; i++)
+            {
+                buffer[i] = client_to_agent_stream_queue[index].front();
+                client_to_agent_stream_queue[index].pop();
+            }
+            
+            printf("Custom agent receive: %ld B in queue %d\n", rv, index);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:135]:
```suggestion
            std::cout << "Custom agent receive: " << rv << " bytes in queue " << index << std::endl;
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    std::unique_lock<std::mutex> lock(agent_to_client_mtx);
+    int32_t index = static_cast<int32_t>(destination_endpoint->get_member<uint32_t>("index"));
+
+    for (size_t i = 0; i < message_length; i++)
+    {
+        agent_to_client_stream_queue[index].emplace(buffer[i]);
+    }
+    
+    transport_rc = eprosima::uxr::TransportRc::ok;
+    printf("Custom agent send: %ld to queue %d\n", message_length, index);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:167]:
```suggestion
    std::cout << "Custom agent send: " << message_length << " bytes to queue " << index << std::endl;
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+    return static_cast<ssize_t>(message_length);
+};
+
+
+// Client custom transport
+extern "C"{
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:174]:
```suggestion
extern "C"
{
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    {
+        (void) errcode;
+
+        int32_t index = *(int32_t*) transport->args;
+
+        std::unique_lock<std::mutex> lock(client_to_agent_mtx);
+
+        std::vector<uint8_t> packet(buf, buf + len);
+        client_to_agent_packet_queue[index].emplace(std::move(packet));
+        printf("Custom client send: %ld to queue %d\n", len, index);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:207]:
```suggestion
        printf("Custom client send: %ld B to queue %d\n", len, index);
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                std::unique_lock<std::mutex> lock(agent_to_client_mtx);
+
+                auto data = agent_to_client_packet_queue[index].front();
+                agent_to_client_packet_queue[index].pop();
+
+                if (data.size() <= len)
+                {
+                    std::copy( data.begin(), data.end(), buf);
+                    rv = data.size();
+                    printf("Custom client received: %ld from queue %d\n", rv, index);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:233]:
```suggestion
                    printf("Custom client received: %ld B from queue %d\n", rv, index);
```

File: test/test/custom_transports/Custom_transports.cpp [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        int32_t index = *(int32_t*) transport->args;
+
+        std::unique_lock<std::mutex> lock(client_to_agent_mtx);
+
+        for (size_t i = 0; i < len; i++)
+        {
+            client_to_agent_stream_queue[index].emplace(buf[i]);
+        }
+
+        printf("Custom client send: %ld to queue %d\n", len, index);
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.cpp:260]:
```suggestion
        printf("Custom client send: %ld B to queue %d\n", len, index);
```

File: test/test/custom_transports/Custom_transports.hpp [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+// Agent custom transports
+extern eprosima::uxr::CustomAgent::InitFunction agent_custom_transport_open;
+extern eprosima::uxr::CustomAgent::FiniFunction agent_custom_transport_close;
+extern eprosima::uxr::CustomAgent::RecvMsgFunction agent_custom_transport_read_stream;
+extern eprosima::uxr::CustomAgent::SendMsgFunction agent_custom_transport_write_stream;
+extern eprosima::uxr::CustomAgent::RecvMsgFunction agent_custom_transport_read_packet;
+extern eprosima::uxr::CustomAgent::SendMsgFunction agent_custom_transport_write_packet;
+
+// Client custom transport
+extern "C"{
```
Comment (jamoralp) [inline test/test/custom_transports/Custom_transports.hpp:16]:
```suggestion
extern "C"
{
```

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/138/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/157/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/139/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/158/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/140/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/159/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/144/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/163/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/145/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/164/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/147/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/166/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/150/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/169/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/152/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/171/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/153/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/172/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/154/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/173/)

File: CMakeLists.txt [hunk 14]
```diff
@@ -29,9 +29,9 @@ endif()
 
 project(microxrcedds VERSION "2.0.0")
 
-set(_client_tag develop)
+set(_client_tag feature/incremental_wait_time)
```
Comment (pablogs9) [inline CMakeLists.txt:32]:
```suggestion
set(_client_tag develop)
```

when https://github.com/eProsima/Micro-XRCE-DDS-Client/pull/204 merged

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/156/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/175/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/159/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/178/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Linux/161/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS%20Manual%20Windows/180/)

**Pull request merged**